[opensprinkler] Improve duration of "open" station

FlorianSW opened the pull request:
This commit improves, how an user can control how long a station will be kept open when switched on.

Fixes: #5619

FlorianSW commented:
> 1. Did we have a default value for the duration before? It looks like it was 68400, is that right? If so, can we implement something like that?

It's still there. If you neither change the value in the thing, not link the nextDuration channel to an item, the default value of the thing will be used as a duration, which is 68400 seconds :)

> 2. Why do you need the `StorageService`? You could also cache that value in the handler, couldn't you?

I talked about this with one user of this binding, and he wished to have the value persisted during restarts as well (which would not be possible when saving in the handler, only, wouldn't it?). He also didn't want to manually setup persistence for this one item only.

J-N-K commented:
> > 1. Why do you need the `StorageService`? You could also cache that value in the handler, couldn't you?
> 
> I talked about this with one user of this binding, and he wished to have the value persisted during restarts as well (which would not be possible when saving in the handler, only, wouldn't it?). He also didn't want to manually setup persistence for this one item only.

I don't think we should do something like that. The persistence is exactly for that purpose and we should not introduce special persistence services for each binding. @openhab/2-x-add-ons-maintainers, WDYT?

FlorianSW commented:
> The persistence is exactly for that purpose

Is it? I kind of don't know. The persistence for me was, until now, a tool for saving historical data, accessing the data from another tool (maybe?) or cloning data to some other service.

However, the communication with a specific device should be handled by the binding, right? And this, from my point of view, looks like a functionality that the device should have. As it does not have it, the binding can (and probably should) implement it.

What, e.g., happens if a user does not have a persistence layer configured (as he didn't need it so far and probably has difficulties setting it up anyhow), changing that the binding holds the value in memory could lead to very strange "errors" and unfulfilled expectations, right?

kaikreuzer commented:
> Is it? I kind of don't know.

Yes, @J-N-K is absolutely right here. This use case is the whole purpose of the ["restoreOnStartup" persistence strategy](https://www.openhab.org/docs/configuration/persistence.html#predefined-strategies).

> as he didn't need it so far and probably has difficulties setting it up anyhow

I actually thought about that since a while already and would like to have a simple persistence configured by default for a plain openHAB installation - persistence is very useful in many different kinds of usecases and thus I agree that we should make it as easy as possible for users to make use of it.

J-N-K commented:
> > The persistence is exactly for that purpose
> 
> Is it? I kind of don't know. The persistence for me was, until now, a tool for saving historical data, accessing the data from another tool (maybe?) or cloning data to some other service.

We have the MapDB persistence service which holds exactly one value per item. Together with the `restoreOnStartup` strategy this is exactly what the user wants.

> What, e.g., happens if a user does not have a persistence layer configured (as he didn't need it so far and probably has difficulties setting it up anyhow), changing that the binding holds the value in memory could lead to very strange "errors" and unfulfilled expectations, right?

What is changing in the binding? This was not available before, if I'm not mistaken.

FlorianSW commented:
Ok, I'm convinced, I've removed the saving part of the binding and added a note to the README, so that users get a hint that a working persistence may be a good idea when using this channel :)

J-N-K commented:
I still have problems understand the conceptual design. Sorry for the hassle,  I really appreciate your work.

If the channel is  not used, the value from the configuration (let's call it `configValue`) is used.
If the user writes a value to the channel (`channelValue`), the `configValue` is "overwritten" and the `channelValue` is used.
A second write to the channel overwrites that again with the new value. Correct?

This looks like a strange mixture of state and configuration.

FlorianSW commented:
> If the channel is not used, the value from the configuration (let's call it configValue) is used.
If the user writes a value to the channel (channelValue), the configValue is "overwritten" and the channelValue is used.
A second write to the channel overwrites that again with the new value. Correct?

Correct :)

> This looks like a strange mixture of state and configuration.

Kind of, the two ways of setting the time for the station to be open are here to support two different use cases (they shouldn't be used together, which wouldn't make much sense).

The first use case is, if you want to open a station for a specific time, and have this value to be static. E.g. if you want to have station1 to be open for 5 minutes when you turn it on through openHAB. In this case, you would choose the thing configuration to configure the time.

The second use case is, if you want to have the possibility to change the time a station is open more frequently. E.g. it's a warm day and you want to turn on the station for 20 minutes and the next time to be open for 10 minutes only. This can be done by the channel state, either by setting it in the app or the UI or by having a rule which sets the time depending on states of other items. In this case, you need to choose the channel in order to set the variable time :)

I hope this makes it more clear what this PR is intended to do :)

J-N-K commented:
I would guess that the correct time calculation is done in a rule. Wouldn't it be better to have a "irrigate" action for rules where you can set the time? For the next opening the standard value is used.

FlorianSW commented:
> I would guess that the correct time calculation is done in a rule. Wouldn't it be better to have a "irrigate" action for rules where you can set the time? For the next opening the standard value is used.

From what I understood from the use case is, that it might be beneficial to have the value persistent, as one would also want to be able to set the time in the app, and probably want to keep the time for multiple "turn on" actions, without needing to set the time again and again :)

J-N-K commented:
I think I got it now. Please give me some time to think about it.

FlorianSW commented:
> I think I got it now. Please give me some time to think about it.

Awesome, you're very welcome if you've further questions or things that need to be done to get the PR merged :)

J-N-K commented:
Just wanted to let you know that this is not forgotten. We are currently discussing methods to let the user change teh thing configuration, but there are some issues to be considered for that. Adding a channel is surely not the right thing to do. See https://github.com/openhab/openhab-core/pull/1004

FlorianSW commented:
While I appreciate that there's a broader discussion on this topic (getting and changing a thing configuration more easily, e.g. with items), this does not seem to be in a state where one would be able to expect something working to come out in the near future (please correct me, if I'm wrong).

So, I would be very interested why this applies here:

> Adding a channel is surely not the right thing to do

There's actually no similar functionality possible with openHAB, and I know at least one user who would actually love to have this feature especially in this binding, so blocking the implementation of it because it may, or may not, be in openHAB core sometimes in the future feels wrong. Especially because, as you mentioned, too, no-one can guarantee that the implementation in openHAB core fits the specific use cases which I explained above :(

J-N-K commented:
Channels are for states/values, not for configuration. The mixing is wrong. 

I would be ok with a channel for the opening time (seems to be similar to the setpoint of a thermostat e.g.). Having a configuration value for that is also ok. But not both at the same time. 

Can you readback the currently set value from the system? If so, just add the channel and remove the configuration parameter. On `REFRESH` read the current value and set the channel to that value. Then a user can easily change that by writing a new value to that channel.

swamiller commented:
I just installed the new opensprinkler binding via the M3 release.  The first thing I noticed was the current "default" duration is 18 HOURS of ontime for a sprinkler zone.  I would love the ability to limit that duration to something more appropriate for a single sprinkler zone, especially for inadvertent manual activation.

FlorianSW commented:
> Channels are for states/values, not for configuration. The mixing is wrong.

I agree with you in general. However, I kind of see this as as something which sets a value to the device, with the only difference, that the device does not support this feature directly, and that's why I implemented it in the binding.

> I would be ok with a channel for the opening time (seems to be similar to the setpoint of a thermostat e.g.). Having a configuration value for that is also ok. But not both at the same time.

Hmm, I think in this case, having the channel only is the alternative which supports all use cases in one or the other way. One would be able to use the channel once and leave it as it is without needing to change it, if they would configure another default value only. That would also fit the use case, @swamiller described.

> Can you readback the currently set value from the system? If so, just add the channel and remove the configuration parameter. On `REFRESH` read the current value and set the channel to that value. Then a user can easily change that by writing a new value to that channel.

I'm not sure what you mean here. What do you mean with "read the value back from the system". If you mean reading the time a sprinkler zone would be turned on the next time, then no, as the time needs to be set when the zone is actually turned on. So, the time would need to be saved in openHAB and remembered when the zone is turned on.

J-N-K commented:
> I'm not sure what you mean here. What do you mean with "read the value back from the system". If you mean reading the time a sprinkler zone would be turned on the next time, then no, as the time needs to be set when the zone is actually turned on. So, the time would need to be saved in openHAB and remembered when the zone is turned on.

Exactly. But this can be done easily with "restoreOnStartup" and e.g. MapDB persistence, which stores exactly the last value. You could give a hint in the documentation about that (i.e. using persistence for the item connected to that channel).

FlorianSW commented:
The last commit removes the thing configuration part, leaving the channel as the only way to change the duration a station will be turned on.

J-N-K commented on bundles/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/internal/handler/OpenSprinklerStationHandler.java:
```diff
@@ -141,7 +157,7 @@ protected State getStationState(int stationId) {
      */
     @Nullable
-    protected State getRemainingWaterTime(int stationId) {
+    private State getRemainingWaterTime(int stationId) {
```
```suggestion
    private @Nullable State getRemainingWaterTime(int stationId) {
```

see https://my.openhab.org/docs/developer/guidelines.html#null-annotations

FlorianSW commented on bundles/org.openhab.binding.opensprinkler/README.md:
```diff
@@ -70,8 +82,8 @@ demo.Things:
 Bridge opensprinkler:http:http [hostname="127.0.0.1", port=81, password="opendoor"] {
-    Thing station 01 [stationIndex=0]
-    Thing station 02 [stationIndex=1]
+    Thing station 01 [stationIndex=0, duration=300]
```
Oops, you're right! :)

J-N-K commented:
Thanks. I really like how this binding is making progress.
